File: src/lib/mailerlite.ts
import axios from "axios";

const MAILERLITE_API_KEY = import.meta.env.VITE_MAILERLITE_API_KEY;
const MAILERLITE_GROUP_ID = import.meta.env.VITE_MAILERLITE_GROUP_ID;

interface MailerLiteFields {
  name?: string;
  last_name?: string;
  [key: string]: string | undefined;
}

interface MailerLiteSubscriber {
  email: string;
  fields?: MailerLiteFields;
  groups?: string[];
}

export async function addSubscriberToMailerLite(
  subscriber: MailerLiteSubscriber
): Promise<boolean> {
  try {
    // Log configuration (without exposing the full API key)
    console.log("MailerLite Configuration:", {
      hasApiKey: !!MAILERLITE_API_KEY,
      apiKeyLength: MAILERLITE_API_KEY?.length || 0,
      groupId: MAILERLITE_GROUP_ID,
    });

    // Log the request payload (without sensitive data)
    console.log("MailerLite Request Payload:", {
      email: subscriber.email,
      fields: subscriber.fields,
      hasGroups: !!subscriber.groups,
      groupsCount: subscriber.groups?.length || 0,
    });

    const response = await axios.post(
      "https://connect.mailerlite.com/api/subscribers",
      {
        ...subscriber,
        groups: [...(subscriber.groups || []), MAILERLITE_GROUP_ID],
      },
      {
        headers: {
          "Content-Type": "application/json",
          Accept: "application/json",
          Authorization: `Bearer ${MAILERLITE_API_KEY}`,
        },
      }
    );

    console.log("MailerLite API Response:", {
      status: response.status,
      statusText: response.statusText,
      data: response.data,
    });

    return response.status === 200 || response.status === 201;
  } catch (error) {
    if (axios.isAxiosError(error)) {
      console.error("MailerLite API Error:", {
        status: error.response?.status,
        statusText: error.response?.statusText,
        data: error.response?.data,
        message: error.message,
      });
    } else {
      console.error("Unknown error adding subscriber to MailerLite:", error);
    }
    return false;
  }
}

File: mailerlite.py
import requests
import json
from typing import List, Dict, Optional

class MailerLiteAPI:
    def __init__(self, api_key: str):
        self.api_key = api_key
        self.base_url = "https://connect.mailerlite.com/api"
        self.headers = {
            "Content-Type": "application/json",
            "Accept": "application/json",
            "Authorization": f"Bearer {api_key}"
        }

    def create_group(self, group_name: str) -> Optional[str]:
        """Create a new group and return its ID"""
        endpoint = f"{self.base_url}/groups"
        data = {"name": group_name}
        
        response = requests.post(endpoint, headers=self.headers, json=data)
        
        if response.status_code in [200, 201]:
            return response.json()["data"]["id"]
        else:
            print(f"Error creating group: {response.text}")
            return None

    def add_subscriber_to_group(self, email: str, group_id: str, fields: Dict = None) -> bool:
        """Add a subscriber to a specific group"""
        endpoint = f"{self.base_url}/subscribers"
        
        data = {
            "email": email,
            "groups": [group_id]
        }
        
        if fields:
            data["fields"] = fields
            
        response = requests.post(endpoint, headers=self.headers, json=data)
        
        if response.status_code in [200, 201]:
            return True
        else:
            print(f"Error adding subscriber: {response.text}")
            return False

    def bulk_add_subscribers_to_group(self, subscribers: List[Dict], group_id: str) -> bool:
        """Bulk add subscribers to a specific group"""
        endpoint = f"{self.base_url}/groups/{group_id}/import-subscribers"
        
        data = {
            "subscribers": subscribers
        }
        
        response = requests.post(endpoint, headers=self.headers, json=data)
        
        if response.status_code == 200:
            return True
        else:
            print(f"Error bulk adding subscribers: {response.text}")
            return False

# Example usage:
if __name__ == "__main__":
    # Initialize the API with your API key
    api_key = "YOUR_API_KEY"
    api = MailerLiteAPI(api_key)
    
    # Create a new group
    group_id = api.create_group("Newsletter Subscribers")
    
    if group_id:
        # Add a single subscriber
        success = api.add_subscriber_to_group(
            "example@email.com",
            group_id,
            fields={"name": "John", "last_name": "Doe"}
        )
        print(f"Single subscriber added: {success}")
        
        # Bulk add subscribers
        subscribers = [
            {
                "email": "subscriber1@email.com",
                "fields": {"name": "Subscriber", "last_name": "One"}
            },
            {
                "email": "subscriber2@email.com",
                "fields": {"name": "Subscriber", "last_name": "Two"}
            }
        ]
        
        success = api.bulk_add_subscribers_to_group(subscribers, group_id)
        print(f"Bulk subscribers added: {success}") 